chunky/src/java/se/llbit/chunky/entity/BeaconBeam.java

```diff
@@ -61,27 +67,135 @@ public BeaconBeam(JsonObject json) {
     this.scale = json.get("scale").asDouble(1);
     this.height = json.get("height").asInt(256);
     this.pose = json.get("pose").object();
+
+    JsonObject materialsList = json.get("beamMaterials").object();
+    for (JsonMember obj : materialsList.members) {
+      BeaconBeamMaterial mat = new BeaconBeamMaterial(BeaconBeamMaterial.DEFAULT_COLOR);
+      mat.loadMaterialProperties(obj.value.object());
+      materials.put(Integer.parseInt(obj.name), mat);
+    }
   }
 
   @Override
-  public Collection<Primitive> primitives(Vector3 offset) {
-    Vector3 allPose = JsonUtil.vec3FromJsonArray(this.pose.get("all"));
-    return primitives(Transform.NONE
-        .translate(-0.5, -0.5, -0.5)
-        .scale(scale)
-        .rotateY(allPose.y) // Ignore x and z rotation because it breaks the beam into segments
-        .translate(0.5, 0.5, 0.5)
-        .translate(position.x + offset.x, position.y + offset.y, position.z + offset.z));
+  public void loadDataFromOctree(Octree octree, BlockPalette palette, Vector3i origin) {
+    int firstColor = BeaconBeamMaterial.DEFAULT_COLOR;
+    boolean foundFirst = false;
+    this.materials.put(0, new BeaconBeamMaterial(BeaconBeamMaterial.DEFAULT_COLOR));
+
+    //Start i at 1 so the first beacon is not checked.
+    //Stop i at 256 even if the beam is taller because the Octree will wrap the coordinates.
+    for (int i = 1; i < height && (i + position.y) < 256; i++) {
+      Material blockMaterial = octree.getMaterial((int)(position.x - origin.x), (int)(position.y + i - origin.y), (int)(position.z - origin.z), palette);
+      int color = colorfromMaterial(blockMaterial);
+      if(color != -1) {
+        if (!foundFirst) {
+          this.materials.put(i, new BeaconBeamMaterial(color));
+          firstColor = color;
+          foundFirst = true;
+        } else {
+          float[] foundColor = new float[3];
+          float[] baseColor = new float[3];
+
+          ColorUtil.getRGBComponents(color, foundColor);
+          ColorUtil.getRGBComponents(firstColor, baseColor);
+
+          int newColor = ColorUtil.getRGB((foundColor[0] + baseColor[0]) / 2.0f, (foundColor[1] + baseColor[1]) / 2.0f, (foundColor[2] + baseColor[2]) / 2.0f);
+          this.materials.put(i, new BeaconBeamMaterial(newColor));
+        }
+      }
+    }
+  }
+
+  private int colorfromMaterial(Material blockMaterial) {
+    if(blockMaterial instanceof Beacon) {
+      return 0xF9FFFE;
+    }
+    if(blockMaterial.name.endsWith("_stained_glass") || blockMaterial.name.endsWith("_stained_glass_pane")) {
```

Comment:
This could probably be refactored to use a regex at some point, e.g. `(minecraft:)?(.+?)_stained_glass(_pane)?` and then it will have the color in the second group and is also future-proof for json model loading (where all block names will be namespaced).